apps/VideoStream/videoStream.py

```diff
@@ -43,50 +41,63 @@ def update(self, address):
 
 class VideoStream:
     """Class for streaming video. """
-   # TODO: Write more complete docstring when distinction between classes is more clear.                                                                    
-
+    # TODO: Write more complete docstring when distinction between classes is more clear.                                                                    
+    # TODO: Add skeletons for additional class methods when functionality of class is made more clear.
+    
     def __init__(self):
         self.status = "down"
         self.port = 1201
         self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        self.connections = Connections(4)
+        self.connections = Connections(2)
+        self.socket.settimeout(1)
 
     def start(self):
         print("Starting VideoStream...")
+        self.listen_thread()
         self.status = "running"
-        self.listen()
 
     def stop(self):
         print("Stopping VideoStream...")
+        threading._shutdown()
+        self.socket.close()
         self.status = "down"
 
     def send_frame(self, frame, address, quality=4):
-        frame = encode_frame(frame)
         encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), quality]
         grey = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
         frame = cv2.imencode('.jpg', grey, encode_param)[1].tostring()
         frame = zlib.compress(frame, -1)
         self.socket.sendto(frame, address)
 
-    def listen(self, port=None):
+    def listen_thread(self):
+        listen = threading.Thread(target = self.listen)
+        listen.daemon = False
+        listen.start()
+
+    def listen(self, port = None):
         if port is None:
             port = self.port
         self.socket.bind(('', port))
         print('Listening on port', port)
         while True:
-            self.connections.cleanup_connections()
-            data, address = self.socket.recvfrom(3)
-            data = data.decode('utf-8')
-            if (data == "get"):
-                self.connections.update(address)
+            self.connections.cleanup()
             for address in self.connections.connections:
                 print(address)
+            try:    
+                data, address = self.socket.recvfrom(3)
+                data = data.decode('utf-8')
+                if (data == "get"):
+                    self.connections.lock.acquire()
+                    self.connections.update(address)
+                    for address in self.connections.connections:
+                        print(address)
+                    self.connections.lock.release()
+                continue
+            except socket.timeout:
+                continue
```

Comment:
The continue statement is needed here because the socket.timeout exception is meant to happen periodically. In the case where videostream has been turned on and is waiting for new connections, the socket.timeout exception gets thrown after one second of waiting.

I can explain better in person, but I basically replaced sleeping with socket.timeout. I did this because the timeout prevents the loop from taking up too much resources while still allowing new connections to be made at any time.

To summarize, socket.recvfrom() is blocking until the bytes are actually received, but there are things that the listen thread needs to do periodically, such as cleaning up the connections and printing current connections. My solution to avoid making another thread was to use the socket.timeout property to re-run the loop at an interval, which is why the loop continues after the socket.timeout exception.

Reply:
Gotcha